packages/neon-api/src/transaction/validator.ts

```diff
@@ -0,0 +1,345 @@
+    const { state } = await this.rpcClient.invokeScript(
+      this.transaction.script
+    );
+    if (state === "FAULT") {
+      return {
+        valid: false,
+        result: {
+          script: {
+            fixed: false,
+            message: "Encountered FAULT when validating script."
+          }
+        }
+      };
+    }
+
+    const intents = new sc.ScriptParser(
+      this.transaction.script
+    ).toScriptParams();
+    const intentsRes: Array<string> = [];
+    await Promise.all(
+      intents.map(intent =>
+        this._validateIntent(intent).then(res =>
+          res ? intentsRes.push(res) : res
+        )
+      )
+    );
+    if (intentsRes.length > 0) {
+      return {
+        valid: false,
+        result: {
+          script: {
+            fixed: false,
+            message: intentsRes.join(";")
+          }
+        }
+      };
+    } else {
+      return {
+        valid: true
+      };
+    }
+  }
+
+  private _validateSystemFee(
+    minimumSystemFee: u.Fixed8,
+    autoFix = false
+  ): ValidationResult {
+    const { systemFee } = this.transaction;
+    if (autoFix && !minimumSystemFee.equals(systemFee)) {
+      this.transaction.systemFee = minimumSystemFee;
+      return {
+        valid: true,
+        result: {
+          systemFee: {
+            fixed: true,
+            prev: systemFee,
+            suggestion: minimumSystemFee
+          }
+        }
+      };
+    } else if (
+      minimumSystemFee.isGreaterThan(systemFee) ||
+      !systemFee.equals(systemFee.ceil())
+    ) {
+      return {
+        valid: false,
+        result: {
+          systemFee: {
+            fixed: false,
+            prev: systemFee,
+            suggestion: minimumSystemFee
+          }
+        }
+      };
+    } else if (minimumSystemFee.isLessThan(systemFee)) {
+      return {
+        valid: true,
+        result: {
+          systemFee: {
+            fixed: false,
+            prev: systemFee,
+            suggestion: minimumSystemFee
+          }
+        }
+      };
+    }
+    return {
+      valid: true
+    };
+  }
+
+  /**
+   * validate systemFee
+   * @param autoFix will automatically fix transaction if specified as true
+   */
+  public async validateSystemFee(autoFix = false): Promise<ValidationResult> {
+    const { script } = this.transaction;
+    const {
+      state,
+      gas_consumed: gasConsumed
+    } = await this.rpcClient.invokeScript(script);
+
+    const minimumSystemFee = new u.Fixed8(parseFloat(gasConsumed))
+      .mul(1e-8)
+      .ceil();
+    const validation = this._validateSystemFee(minimumSystemFee, autoFix);
+
+    if (state === "FAULT") {
+      return {
+        valid: false,
+        result: {
+          systemFee: {
+            ...{
+              fixed: false,
+              message:
+                "Cannot get precise systemFee as script execution on node reports FAULT."
+            },
+            ...validation.result?.systemFee
```

Comment:
I thought we discussed and this suggestion should only appear when the suggestion is smaller than the currently attached fee.